### PR TITLE
End of night shutdown improvements.

### LINKED ISSUE #390: End of night shutdown
At the end of the night the scheduler is not finding any valid observations, which is okay, but the system sort of treats this as an error and shuts down. This triggers a  "error" shut down that stops the state machine, meaning it won't come up automatically the next night. Output from logs is here:

```
I0118 15:41:30.891 messaging.py:122     PANCHAT Weather is good and it is dark. Something must have gone wrong. Stopping loop
I0118 15:41:30.892 machine.py:171       Stopping POCS states
```

See also #373

### PR DESCRIPTION
**Pull request by wtgee**
A number of fixes and changes to account for situation when no valid
observations are found but we don't want to shut down the state machine.

* If scheduler has a list of `observations` then it `has_valid_observations`.
	which is True when starting.

Process:
* `scheduling` state calls `observatory.get_observation()`
* `get_observation` first checks if `has_valid_observations is False`
	if so it will reread the field list.
* `get_observation` either returns an observation or, if none found,
	calls `clear_available_observations`, making `has_valid_observations = False`.
* If no observation is returned to `scheduling` state, send to `parking` (then to `sleeping`).
* In `sleeping`, if `has_valid_observations is False` then sleep for an hour
	then try again (or shut down if `run_once` is specified).
  If `has_valid_observations is True` and it is safe and dark, shut down because
  	of unknown error.

Other changes:
* Remove `reread_fields_file` from dispatch scheduler as it is handled at
the `observatory` level by `observatory.get_observation`.
* Misc small fixes

> ⚠️ Note: I think I've gotten all the logic correct so that Bad Things™ won't happen. I think. :)

Closes #390

### CODE REVIEW

File: pocs/state/states/default/sleeping.py [hunk 1]
```diff
@@ -4,11 +4,17 @@ def on_enter(event_data):
     pocs.next_state = 'ready'
 
     # If it is dark and safe we shouldn't be in sleeping state
-    if pocs.is_dark() and pocs.is_safe():
+    if pocs.observatory.scheduler.has_valid_observations and \
```
Comment (jamessynge) [inline pocs/state/states/default/sleeping.py]: pep8 strongly prefers wrapping with parens around expressions rather than using a backslash.

File: pocs/state/states/default/sleeping.py [hunk 2]
```diff
@@ -4,11 +4,17 @@ def on_enter(event_data):
...
+            pocs.is_dark() and pocs.is_safe():
         if pocs.should_retry is False:
```
Comment (jamessynge) [inline pocs/state/states/default/sleeping.py]: Property should_retry is very problematic. It looks like a property, but accessing it modifies its value!

File: pocs/tests/test_dispatch_scheduler.py [hunk 3]
```diff
@@ -119,11 +119,12 @@ def test_get_observation_reread(field_list, observer, temp_file, constraints):
         f.write(yaml.dump([{
             'name': 'New Name',
             'position': '20h00m43.7135s +22d42m39.0645s',
-            'priority': 50
+            'priority': 5000
```
Comment (jamessynge) [inline pocs/tests/test_dispatch_scheduler.py]: Out of curiosity, why the change in priority?

File: pocs/observatory.py [hunk 4]
```diff
@@ -184,9 +184,16 @@ def get_observation(self, *args, **kwargs):
         """
 
         self.logger.debug("Getting observation for observatory")
+
+        # If observation list is empty or a reread is requested
+        if self.scheduler.has_valid_observations is False or \
```
Comment (jamessynge) [inline pocs/observatory.py]: pep8 strongly prefers wrapping with parens around expressions rather than using a backslash.

File: pocs/state/states/default/sleeping.py (hunk 2 above)
Comment (wtgee) [reply pocs/state/states/default/sleeping.py]: Yes, thanks for calling this out. It's also problematic in that it is never reset without stopping and starting whole system. So if it retries once per night by the third night it is out of retry attempts no matter what. Will fix.

**Comment by wtgee** [comment]
I've removed a whole test in `test_pocs.py` that was hanging. Looking at the test I can't figure out how it was any different from the test below it. Based off of the name of the test, it was designed to be testing some of the interrupt code that a student added about a year ago (via Huntsman) and which has been removed during repository merge and cleanup because it needs some independent work.

Feel weird about just removing a test but I think it is the right course of action.

File: pocs/observatory.py [hunk 5]
```diff
@@ -186,8 +186,8 @@ def get_observation(self, *args, **kwargs):
         self.logger.debug("Getting observation for observatory")
 
         # If observation list is empty or a reread is requested
-        if self.scheduler.has_valid_observations is False or \
-                kwargs.get('reread_fields_file', False):
+        if (self.scheduler.has_valid_observations is False or
+                kwargs.get('reread_fields_file', False)):
             self.scheduler.read_field_list()
 
         self.scheduler.get_observation(*args, **kwargs)
```
Comment (jamessynge) [inline pocs/observatory.py]: This looks like it should return something. Why not return what get_observation returns?

File: pocs/state/machine.py [hunk 6]
```diff
@@ -129,7 +130,8 @@ def run(self, exit_when_done=False, run_once=False):
                 try:
                     state_changed = self.goto_next_state()
                 except Exception as e:
-                    self.logger.warning("Problem going to next state, exiting loop [{}]".format(e))
+                    self.logger.warning("Problem going from {} to {}, exiting loop [{!r}]".format(
```
Comment (jamessynge) [inline pocs/state/machine.py]: Fine for now, but it would be nice to have a recovery mechanism, so that the system automatically tries to get to mount-parked/dome-closed/state-sleeping if things go bad, and then resumes... or waits until the next noon to restart.

File: pocs/state/states/default/parked.py [hunk 7]
```diff
@@ -4,5 +4,25 @@ def on_enter(event_data):
     pocs = event_data.model
     pocs.say("I'm parked now. Phew.")
 
-    pocs.say("Cleaning up for the night!")
-    pocs.next_state = 'housekeeping'
+    has_valid_observations = pocs.observatory.scheduler.has_valid_observations
+
+    if (has_valid_observations and pocs.is_safe() is False):
```
Comment (jamessynge) [inline pocs/state/states/default/parked.py]: Parentheses not needed here.

File: pocs/state/states/default/parked.py [hunk 8]
```diff
@@ -4,5 +4,25 @@ def on_enter(event_data):
...
+        pocs.say("Cleaning up for the night!")
+        pocs.next_state = 'housekeeping'
+    elif (has_valid_observations and pocs.is_safe()):
```
Comment (jamessynge) [inline pocs/state/states/default/parked.py]: Excess parens.

File: pocs/state/states/default/parked.py (hunk 7 above)
Comment (jamessynge) [inline pocs/state/states/default/parked.py]: Perhaps:
```python
if has_valid_observations:
  if not pocs.is_safe():
    ...
  else:  # pocs.is_safe
    ...
else: # not has_valid_observations
  ...
```

File: pocs/observatory.py (hunk 5 above)
Comment (wtgee) [reply pocs/observatory.py]: Good idea to make it explicit, but `get_observation` actually also sets `current_observation`, which is then returned. Should maybe think about renaming this.

File: pocs/state/machine.py (hunk 6 above)
Comment (wtgee) [reply pocs/state/machine.py]: The `state_changed` is `False` it will do as you describe, which should be what happens under normal operations. We should probably do as you describe although I would want to do more testing and focus on it. Created #419 to address.
